crates/transaction-pool/src/maintain/mod.rs

```diff
@@ -0,0 +1,363 @@
+pub struct MaintainPoolConfig {
+    /// Maximum (reorg) depth we handle when updating the transaction pool: `new.number -
+    /// last_seen.number`
+    ///
+    /// Default: 64 (2 epochs)
+    pub max_update_depth: u64,
+    /// Maximum number of accounts to reload from state at once when updating the transaction pool.
+    ///
+    /// Default: 100
+    pub max_reload_accounts: usize,
+
+    /// Maximum amount of time non-executable, non local transactions are queued.
+    /// Default: 3 hours
+    pub max_tx_lifetime: Duration,
+
+    /// Apply no exemptions to the locally received transactions.
+    ///
+    /// This includes:
+    ///   - no price exemptions
+    ///   - no eviction exemptions
+    pub no_local_exemptions: bool,
+}
+
+impl Default for MaintainPoolConfig {
+    fn default() -> Self {
+        Self {
+            max_update_depth: 64,
+            max_reload_accounts: 100,
+            max_tx_lifetime: MAX_QUEUED_TRANSACTION_LIFETIME,
+            no_local_exemptions: false,
+        }
+    }
+}
+
+/// Settings for local transaction backup task
+#[derive(Debug, Clone, Default)]
+pub struct LocalTransactionBackupConfig {
+    /// Path to transactions backup file
+    pub transactions_path: Option<PathBuf>,
+}
+
+impl LocalTransactionBackupConfig {
+    /// Receive path to transactions backup and return initialized config
+    pub const fn with_local_txs_backup(transactions_path: PathBuf) -> Self {
+        Self { transactions_path: Some(transactions_path) }
+    }
+}
+
+/// Returns a spawnable future for maintaining the state of the transaction pool.
+pub fn maintain_transaction_pool_future<N, Client, P, St, Tasks>(
+    client: Client,
+    pool: P,
+    events: St,
+    task_spawner: Tasks,
+    config: MaintainPoolConfig,
+) -> BoxFuture<'static, ()>
+where
+    N: NodePrimitives,
+    Client: StateProviderFactory + BlockReaderIdExt + ChainSpecProvider + Clone + 'static,
+    P: TransactionPoolExt<Transaction: PoolTransaction<Consensus = N::SignedTx>> + 'static,
+    St: Stream<Item = CanonStateNotification<N>> + Send + Unpin + 'static,
+    Tasks: TaskSpawner + 'static,
+{
+    async move {
+        maintain_transaction_pool(client, pool, events, task_spawner, config).await;
+    }
+    .boxed()
+}
+
+/// Maintains the state of the transaction pool by handling new blocks and reorgs.
+///
+/// This listens for any new blocks and reorgs and updates the transaction pool's state accordingly
+pub async fn maintain_transaction_pool<N, Client, P, St, Tasks>(
+    client: Client,
+    pool: P,
+    mut events: St,
+    task_spawner: Tasks,
+    config: MaintainPoolConfig,
+) where
+    N: NodePrimitives,
+    Client: StateProviderFactory + BlockReaderIdExt + ChainSpecProvider + Clone + 'static,
+    P: TransactionPoolExt<Transaction: PoolTransaction<Consensus = N::SignedTx>> + 'static,
+    St: Stream<Item = CanonStateNotification<N>> + Send + Unpin + 'static,
+    Tasks: TaskSpawner + 'static,
+{
```

Comment:
I like the direction here, but I think we can even go a step further,

for example, let's see the customization usecase is:
for every event we also need to do x, or want to change how the event is processed entirely,

then it would be great if this entire type would be its own future that represents the update loop

Reply:
still drift monitor and canon event processor are not configurable, will ping you when ready

Reply:
@mattsse it is ready for another pass, ptal